### PR TITLE
[Optimization] Support multimodal runner for image/video feature processing

### PR DESCRIPTION
**Pull request by xiaoxiaohehe001**
## Motivation
支持多模态 runner 中图像/视频特征处理流程，增强 GPU Model Runner 对预编码多模态特征的处理能力。

### fastdeploy/worker/gpu_model_runner.py
- 新增对 `image_feature_urls` 预编码图像特征的处理逻辑，支持直接传入已编码的 image embedding，跳过 vision encoder 计算
- 新增 `image_grid_thws` 和 `video_features` / `video_grid_thws` 的传递与管理
- Prefill 阶段增加多模态 attention mask offsets 的计算与设置（`attn_mask_offsets`、`decode_states`）
- 调用 `update_attn_mask_offsets` 在 forward 前更新 attention mask
- 将 `attn_mask_offsets` 传入 forward meta，供模型推理使用
- 对 prefill 请求按 `idx` 排序，确保处理顺序一致性
- Forward 结束后主动清空 `image_features` / `video_features` 等中间状态，防止内存泄漏

### fastdeploy/worker/input_batch.py
- 新增 `image_grid_thws`、`video_features`、`video_grid_thws`、`video_infinity_scales` 字段
- 新增 `decode_states`、`attn_mask_offsets`、`attn_mask_offsets_full` tensor 初始化
- 在 `swap_data`、`reset`、`resize` 等操作中补齐新增字段的维护逻辑
- 补充 `generated_modality` 在 swap 和 reset 中的处理（之前遗漏）

### fastdeploy/engine/sched/resource_manager_v1.py
- 移除 Ernie5 架构下多模态请求的特殊调度限制（`get_enough_request`），统一调度逻辑

## Checklist

- [ ] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [ ] Format your code, run `pre-commit` before commit.
- [ ] Add unit tests. Please write the reason in this PR if no unit tests.
- [ ] Provide accuracy results.
- [ ] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: fastdeploy/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -791,6 +834,22 @@ def insert_tasks_v1(self, req_dicts: List[Request], num_running_requests: int =
...
+
+                if self.enable_mm:
+                    self.share_inputs["decode_states"][idx, 0] = 0
+                    inputs = request.multimodal_inputs
+                    # mm attention_mask
+                    attn_offset_len = prefill_end_index - prefill_start_index
+                    if inputs.get("attention_mask_offset", None) is None:
+                        attention_mask_offset_slice = np.arange(prefill_start_index, prefill_end_index, dtype=np.int32)
+                    else:
+                        attention_mask_offset_slice = np.asarray(
```
Comment (freeliuzc) [inline fastdeploy/worker/gpu_model_runner.py:846]: 这里会引入同步，用 async_set_value 改写

File: fastdeploy/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -626,6 +632,42 @@ def _process_mm_features(self, request_list: List[Request]):
...
+                                )
+                        image_features_gpu = [vf.cuda() for vf in image_feature]
+                        image_embeds = paddle.concat(image_features_gpu, axis=0)
+                        multi_vision_inputs["image_features"].append(image_embeds)
+                        logger.info("Enable Encode image embedding.")
+                    else:
+                        multi_vision_inputs["image_features"].extend(image_feature)
+                        logger.info("Disable Encode image embedding.")
+
+        self.share_inputs["image_features"] = multi_vision_inputs["image_features"]
```
Comment (freeliuzc) [inline fastdeploy/worker/gpu_model_runner.py:662]: 这里可以直接访问，不用判断吗？

File: fastdeploy/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -626,6 +632,42 @@ def _process_mm_features(self, request_list: List[Request]):
...
+        if len(multi_vision_inputs["image_features"]) > 0:
+            if (
+                isinstance(multi_vision_inputs["image_features"][0], paddle.Tensor)
+                and len(multi_vision_inputs["image_features"][0].shape) == 2
+            ):
+                self.share_inputs["image_features"] = paddle.concat(multi_vision_inputs["image_features"], axis=0)
+        self.share_inputs["image_grid_thws"] = multi_vision_inputs["image_grid_thws"]
```
Comment (freeliuzc) [inline fastdeploy/worker/gpu_model_runner.py:669]: 同上

File: fastdeploy/worker/input_batch.py [hunk 4]
```diff
@@ -342,7 +346,26 @@ def init_share_inputs(self):
...
+                dtype="int32",
+            )
+            self.attn_mask_offsets = paddle.full(
+                shape=[self.scheduler_config.max_num_seqs * self.model_config.max_model_len],
+                fill_value=-1,
+                dtype="int32",
+            )
+            self.attn_mask_offsets_full = paddle.full(
+                [self.scheduler_config.max_num_seqs, self.model_config.max_model_len], -1, dtype="int32"
+            )
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/input_batch.py:368]: 不需要修改。 update_attn_mask_offsets.cu 中算子每次调用都会自己创建一个新的临时tensor，并且 attn_mask_offsets 写回是通过copy，而非shape必须一致，混淆了“算子内部创建的临时输出tensor"和"预分配的att mask fses bufer"之间的关系。预分配的buffer是接收端，算子输出是发送端，通过copy_桥接，不要求 shape完全一致，只要dst足够大即可。当前的分配策略在调度器约束下是安全的。

File: fastdeploy/worker/input_batch.py [hunk 5]
```diff
@@ -409,6 +432,7 @@ def swap_data(tensor, idx1, idx2):
         swap_data(self.ori_seq_lens_encoder, i1, i2)
         swap_data(self.system_lens, i1, i2)
         swap_data(self.system_ids, i1, i2)
+        swap_data(self.generated_modality, i1, i2)
         swap_data(self.enable_thinking, i1, i2)
         swap_data(self.max_think_lens, i1, i2)
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/input_batch.py:437]: 已修改

File: fastdeploy/worker/gpu_model_runner.py [hunk 6]
```diff
@@ -626,6 +631,38 @@ def _process_mm_features(self, request_list: List[Request]):
                             prefill_end_index=request.prefill_end_index,
                         )
                     )
+            if inputs.get("image_feature_urls", None) is not None and len(inputs["image_feature_urls"]) > 0:
+                multi_vision_inputs["image_grid_thws"].extend(
+                    inputs["image_grid_thws"][request.image_start : request.image_end]
+                )
+                image_feature = inputs["image_features"][request.image_start : request.image_end]
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py:657]: 不需要修改，inputs["image_features"] 返回的list, 每个URL对应一个元素（按图片粒度），用图片计数索引[image_start:image_end] 去切这些 per-image 的 list 没问题

File: fastdeploy/worker/gpu_model_runner.py [hunk 7]
```diff
@@ -626,6 +631,38 @@ def _process_mm_features(self, request_list: List[Request]):
...
+                        # Enable encode vision_embedding
+                        for image_feature_tensor in image_feature:
+                            if image_feature_tensor.shape[1] != self.fd_config.model_config.hidden_size:
+                                logger.error(
+                                    f"Shape mismatch: expected shape={self.fd_config.model_config.hidden_size}, \
+                                        but got {image_feature_tensor.shape}"
+                                )
+                        image_features_gpu = [vf.cuda() for vf in image_feature]
+                        image_embeds = paddle.concat(image_features_gpu, axis=0)
+                        multi_vision_inputs["image_features"].append(image_embeds)
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py:670]: 暂不修改，这个分支处理的是已经下载好的预编码特征 image_feature_urls路径)，shape不匹配属于上游数据错误，在实际生产中极少出现

File: fastdeploy/worker/gpu_model_runner.py [hunk 8]
```diff
@@ -791,6 +829,18 @@ def insert_tasks_v1(self, req_dicts: List[Request], num_running_requests: int =
...
+
+                if self.enable_mm:
+                    self.share_inputs["decode_states"][idx, 0] = 0
+                    inputs = request.multimodal_inputs
+                    # mm attention_mask
+                    if inputs.get("attention_mask_offset", None) is None:
+                        inputs["attention_mask_offset"] = np.arange(self.model_config.max_model_len).tolist()
+                    attn_offset_len = prefill_end_index - prefill_start_index
+                    self.share_inputs["attn_mask_offsets_full"][idx, 0:attn_offset_len] = paddle.to_tensor(
+                        inputs["attention_mask_offset"][prefill_start_index:prefill_end_index], dtype="int32"
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py]: 已修改

File: fastdeploy/worker/gpu_model_runner.py [hunk 9]
```diff
@@ -2241,6 +2305,16 @@ def _preprocess(
         model_inputs["generated_modality"] = self.share_inputs["generated_modality"]
         if self.enable_mm:
             model_inputs["image_features"] = self.share_inputs["image_features"]
+            model_inputs["decode_states"] = self.share_inputs["decode_states"]
+            model_inputs["image_grid_thws"] = self.share_inputs.get("image_grid_thws", None)
+            model_inputs["video_features"] = self.share_inputs["video_features"]
+            model_inputs["video_grid_thws"] = self.share_inputs.get("video_grid_thws", None)
+
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py]: 已修改

File: fastdeploy/worker/gpu_model_runner.py [hunk 10]
```diff
@@ -86,6 +86,7 @@
         set_data_ipc,
         unset_data_ipc,
         get_position_ids_and_mask_encoder_batch,
+        update_attn_mask_offsets,
     )
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py:90]: 已修改

File: fastdeploy/worker/input_batch.py [hunk 11]
```diff
@@ -662,7 +687,13 @@ def reset_share_inputs(self):
                     dtype="float32",
                 )
                 self.image_features = None
+                self.image_grid_thws = None
                 self.image_features_list = None
+                self.video_features = None
+                self.video_grid_thws = None
+                self.video_infinity_scales = None
+                fill_paddle_tensor(self, "attn_mask_offsets", -1)
+                fill_paddle_tensor(self, "attn_mask_offsets_full", -1)
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/input_batch.py:699]: 已修改

File: fastdeploy/worker/input_batch.py [hunk 12]
```diff
@@ -892,6 +927,7 @@ def swap_data(tensor, idx1, idx2):
         swap_data(self.mask_rollback, i1, i2)
         swap_data(self.recompute_token_num, i1, i2)
         if self.enable_mm:
+            swap_data(self.attn_mask_offsets, i1, i2)
             swap_data(self.attn_mask_offsets_full, i1, i2)
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/input_batch.py:935]: 不需要修改

File: fastdeploy/worker/gpu_model_runner.py [hunk 13]
```diff
@@ -626,6 +632,42 @@ def _process_mm_features(self, request_list: List[Request]):
...
                     )
+            if (
+                inputs is not None
+                and inputs.get("image_feature_urls", None) is not None
+                and len(inputs["image_feature_urls"]) > 0
+            ):
+                multi_vision_inputs["image_grid_thws"].extend(
+                    inputs["image_grid_thws"][request.image_start : request.image_end]
+                )
+                image_feature = inputs["image_features"][request.image_start : request.image_end]
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py:657]: 不需要修改，使用image_feature_urls 路径，须同时提供 image_grid_thws

File: fastdeploy/worker/input_batch.py [hunk 14]
```diff
@@ -342,7 +346,26 @@ def init_share_inputs(self):
...
+            self.video_infinity_scales = None
+
+            decode_states_len = self.speculative_config.num_speculative_tokens + 1 if self.speculative_decoding else 1
+            self.decode_states = paddle.full(
+                [self.scheduler_config.max_num_seqs, decode_states_len],
+                -1,
+                dtype="int32",
+            )
+            self.attn_mask_offsets = paddle.full(
+                shape=[self.scheduler_config.max_num_seqs * self.model_config.max_model_len],
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/input_batch.py:362]: 不需要修改，上面已回答

File: fastdeploy/worker/input_batch.py [hunk 15]
```diff
@@ -892,6 +928,8 @@ def swap_data(tensor, idx1, idx2):
         swap_data(self.mask_rollback, i1, i2)
         swap_data(self.recompute_token_num, i1, i2)
         if self.enable_mm:
+            swap_data(self.decode_states, i1, i2)
+            swap_data(self.attn_mask_offsets, i1, i2)
             swap_data(self.attn_mask_offsets_full, i1, i2)
             swap_data(self.attn_mask_offsets_decoder, i1, i2)
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/input_batch.py:936]: 不需要修改

File: fastdeploy/worker/gpu_model_runner.py [hunk 16]
```diff
@@ -187,6 +188,9 @@ def __init__(
             else:
                 self.encoder_cache = None
 
+            # Note(Zhengshifeng) init video cache for VL model
+            self.video_cache = {}
+
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py:194]: 不需要修改

File: fastdeploy/worker/gpu_model_runner.py [hunk 17]
```diff
@@ -1375,6 +1447,7 @@ def initialize_forward_meta(self, is_dummy_or_profile_run=False):
             kv_batch_ids=self.share_inputs["kv_batch_ids"],
             kv_tile_ids_per_batch=self.share_inputs["kv_tile_ids_per_batch"],
             kv_num_blocks_x_cpu=self.share_inputs["kv_num_blocks_x_cpu"],
+            attn_mask_offsets=self.share_inputs["attn_mask_offsets"] if self.enable_mm else None,
             routing_replay_table=routing_replay_table,
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py:1475]: 不需要修改

File: fastdeploy/worker/gpu_model_runner.py (hunk 1 above)
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py:869]: 已修改

File: fastdeploy/worker/gpu_model_runner.py (hunk 2 above)
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py:676]: 前置有初始化为空 list

File: fastdeploy/worker/gpu_model_runner.py (hunk 3 above)
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py:683]: 前置有初始化为空 list

File: fastdeploy/worker/gpu_model_runner.py [hunk 18]
```diff
@@ -812,6 +855,23 @@ def insert_tasks_v1(self, req_dicts: BatchRequest, num_running_requests: int = N
                 prefill_start_index = request.prefill_start_index
                 prefill_end_index = request.prefill_end_index
                 length = prefill_end_index - prefill_start_index
+
```
Comment (K11OntheBoat) [inline fastdeploy/worker/gpu_model_runner.py:860]: self.enable_mm 的语意是否和 enable_mm_runtime 有冲突？self.enable_mm 现在仍然代表“具有多模能力”的模型吗？runner里的一些if else判断，是否应该按照enable_mm_runtime来判断？

Comment (K11OntheBoat) [reply fastdeploy/worker/gpu_model_runner.py:860]: 比如说，我部署EB5模型，环境变量用FD runner，起服务加上--deploy-modality 'text' \， 这里的self.enable_mm 内部的代码是否是多余的？

Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py:860]: 在runner中，self.enable_mm = self.fd_config.enable_mm_runtime 是同一值

File: fastdeploy/worker/gpu_model_runner.py [hunk 19]
```diff
@@ -756,6 +798,7 @@ def insert_tasks_v1(self, req_dicts: BatchRequest, num_running_requests: int = N
             "position_ids_offset": [0],
             "max_tokens_lst": [],
         }
+        req_dicts = sorted(req_dicts, key=lambda r: r.idx)
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/gpu_model_runner.py]: 已修改

File: fastdeploy/worker/input_batch.py [hunk 20]
```diff
@@ -451,6 +475,8 @@ def swap_data(tensor, idx1, idx2):
                     self.image_features_list[i1],
                 )
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/input_batch.py:476]: 不需修改

File: fastdeploy/worker/input_batch.py [hunk 21]
```diff
@@ -342,7 +346,26 @@ def init_share_inputs(self):
...
+
+            decode_states_len = self.speculative_config.num_speculative_tokens + 1 if self.speculative_decoding else 1
+            self.decode_states = paddle.full(
+                [self.scheduler_config.max_num_seqs, decode_states_len],
+                -1,
+                dtype="int32",
+            )
+            self.attn_mask_offsets = paddle.full(
+                shape=[self.scheduler_config.max_num_seqs * self.model_config.max_model_len],
+                fill_value=-1,
```
[the remark this answers is not in the document]
Comment (xiaoxiaohehe001) [reply fastdeploy/worker/input_batch.py:363]: 不需修改

**Review by TBD1 — APPROVED** [review]

**Review by K11OntheBoat — APPROVED** [review]
LGTM

**Review by zyyzghb — APPROVED** [review]

**Review by freeliuzc — APPROVED** [review]
